File: compiler/bench/linsearch.c
// expected: 49950000
// Linear search benchmark - search 100K times in unsorted array

#include <stdio.h>

#define SIZE 1000
#define ITERS 100000

int arr[SIZE];

int lsearch(int val) {
    for (int i = 0; i < SIZE; i++) {
        if (arr[i] == val) return i;
    }
    return -1;
}

void init_arr(void) {
    // Shuffle pattern: store (i*7) mod SIZE at index i
    for (int i = 0; i < SIZE; i++) {
        arr[i] = (i * 7) % SIZE;
    }
}

long bench(void) {
    long sum = 0;
    for (int i = 0; i < ITERS; i++) {
        sum += lsearch(i % SIZE);
    }
    return sum;
}

int main(void) {
    init_arr();
    printf("%ld\n", bench());
    return 0;
}

File: compiler/bench/linsearch.fs
\ expected: 49950000
\ Linear search benchmark - search 100K times in unsorted array

1000 constant SIZE
100000 constant ITERS
create arr SIZE cells allot

: arr@ ( i -- n ) cells arr + @ ;
: arr! ( n i -- ) cells arr + ! ;

: lsearch ( val -- idx|-1 )
  SIZE 0 do
    dup i arr@ = if drop i unloop exit then
  loop
  drop -1 ;

: init-arr ( -- )
  \ Shuffle pattern: store (i*7) mod SIZE at index i
  SIZE 0 do
    i 7 * SIZE mod i arr!
  loop ;

: bench ( -- sum )
  0
  ITERS 0 do
    i SIZE mod lsearch +
  loop ;

: main
  init-arr
  bench . cr ;
